src/GafferImage/RankFilter.cpp

```diff
@@ -181,6 +182,547 @@ Imath::Box2i RankFilter::computeDataWindow( const Gaffer::Context *context, cons
+// the update for most rows will just be a check that everything is still valid, with only the newly updated
+// row requiring a rebalance. There may be a small optimization here where we track that the heap is still
+// valid, and only the new row needs to be updated, but I suspect calling update on each element even though
+// it doesn't need to move is pretty cheap.
+//
+// The result of all this is an extremely efficient algorithm. The definite majority of cost is just in
+// sorting each row. It seems very likely that this approach is competitive with the best of median sorts
+// in the literature - I should probably document it a bit more cleanly and publish that somewhere. If we
+// really wanted to compete with bleeding edge, the next step would probably be to switch the sort of
+// individual rows from std::sort to a bleeding edge sorting algorithm ( like sorting networks for small sizes ).
+
+class RankMedianBuffer
+{
+public:
+	inline RankMedianBuffer( const V2i &size ) :
+		m_size( size ), m_sortedRows( size.x * size.y ), m_splits( size.y ), m_splitValue( 0 )
+	{
+		// We need to initialize the storage for everything
+		m_minHeap.reserve( size.y );
+		m_minHeapHandles.reserve( size.y );
+		m_maxHeap.reserve( size.y );
+		m_maxHeapHandles.reserve( size.y );
+
+		// And the heaps need to have an entry added for each row ( the priority values are not meaningful,
+		// they will all be updated before we use the heaps ).
+		for( int i = 0; i < size.y; i++ )
+		{
+			m_minHeapHandles.push_back( m_minHeap.push( std::make_pair( 0.0f, i ) ) );
+			m_maxHeapHandles.push_back( m_maxHeap.push( std::make_pair( 0.0f, i ) ) );
+		}
+	}
+
+	inline void setRow( int index, Sampler &sampler, const Box2i &visitBound )
+	{
+		// Find the chunk of pixels in m_sortedRows corresponding to this row
+		float *currentRow = &m_sortedRows[ m_size.x * index ];
+
+		// Grab the row of pixels into the buffer
+		float *writePos = currentRow;
+		sampler.visitPixels( visitBound,
+			[&writePos] ( float v, int x, int y )
+			{
+				if( std::isnan( v ) )
+				{
+					v = -infinity;
+				}
+				*writePos = v;
+				writePos++;
+			}
+		);
+
+		// Sort the new row
+		//
+		// If we wanted to make this competitive with bleeding edge median filters, we would probably
+		// need a faster sort here. Sorting networks could work for small sizes ... I'm also curious how
+		// djbsort would perform.
+		std::sort( currentRow, currentRow + m_size.x );
+
+		// Update m_splits for this row to preserve the invariant - it needs to be set so that
+		// currentRow[i] < m_splitValue iff i < m_splits[i]
+		//
+		// This means that the comparison invariant is preserved, but the number of elements in the
+		// lower set may now be wrong - this will be rebalanced when currentResult() is called.
+		int i;
+		for( i = 0; i < m_size.x; i++ )
+		{
+			if( currentRow[i] >= m_splitValue )
+			{
+				break;
+			}
+		}
```

Comment:
Good point. This was 0.02 second reduction in my test - negligible compared to the original runtime, but about 3% of the new runtime.

I do wonder whether this might not be a win for more realistic sizes of median filter - with a radius less than 10, a linear scan with good branch prediction might beat a more sophisticated approach, but I don't have a test for that currently, so I'll just go with lower_bound for now.